codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/transformers/RefactorConstrainedMemberType.kt

```diff
@@ -0,0 +1,195 @@
+package software.amazon.smithy.rust.codegen.server.smithy.transformers
+
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.AbstractShapeBuilder
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.rust.codegen.core.smithy.DirectedWalker
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.server.smithy.allConstraintTraits
+import software.amazon.smithy.rust.codegen.server.smithy.hasConstraintTrait
+import software.amazon.smithy.rust.codegen.server.smithy.traits.RefactoredStructureTrait
+import software.amazon.smithy.utils.ToSmithyBuilder
+import java.lang.IllegalStateException
+import java.util.*
+
+object RefactorConstrainedMemberType {
+    private data class MemberTransformation(
+        val newShape: Shape,
+        val memberToChange: MemberShape,
+        val traitsToKeep: List<Trait>,
+    )
+
+    /**
+     * Transforms all member shapes that have constraints on them into equivalent
+     * non-constrained member shapes.
+     */
+    fun transform(model: Model): Model {
+        val existingRefactoredNames = HashSet<ShapeId>()
+
+        val transformations = model.operationShapes
+            .flatMap { operation ->
+                listOfNotNull(operation.input.orNull(), operation.output.orNull())
+            }
+            .map { model.expectShape(it) }
+            .flatMap {
+                val walker = DirectedWalker(model)
+                walker.walkShapes(it)
+            }
+            .filterIsInstance(StructureShape::class.java)
```

Comment:
All this does is getting reachable structure shapes. We can just use `model.structureShapes` here; the plugin takes care of not generating shapes that are not reachable from an operation.